sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/Utility.java

```diff
@@ -243,6 +243,18 @@ public static OffsetDateTime parseDate(String dateString) {
                     throw LOGGER.logExceptionAsError(new RuntimeException("I/O errors occurs. Error details: "
                         + e.getMessage()));
                 }
+            })
+            .doFirst(() -> {
+                /*
+                If the request needs to be retried, the flux will be resubscribed to. The stream and counter must be
+                reset in order to correctly return the same data again.
+                 */
+                currentTotalLength[0] = 0;
+                try {
+                    data.reset();
```

Comment:
Not all input streams support `reset()`. Some may throw IOException as specified in [javadoc](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#reset--). Also, if the input stream does support resetting the position, this will reset to the last marked position that may be different from where the user wanted us to read the stream from.

Reply:
Yes.

The most common `InputStream` in the upload case, FileInputStream (FIS), does not support `reset`. Users might have to wrap FIS in `BufferedInputStream` (BIS) and give it as input to upload API. But BIS comes with extra allocation cost as worse as the size of the file.

Few options I could think of are:

1. If we choose reset path, then make it clear in the java-doc that it is required to have "reset to the start" support for retry. Since `FileInputStream` is the most common and does not support reset out of the box, having a code sample/snippet showing how to create `InputStream` from `FileChannel` may help. `FileChannel` backing File does support setting read position.

2. another option is to take a `Supplier<InputStream>` which means, each subscription will invoke Supplier to get a new InputStream (hence read-position 0 for each instance), But this is a API change and needs to discuss and get approved, there are questions like the pattern for disposing of the obtained IS.

Reply:
Yes, I have added docs to all the apis which call this method indicating that the stream must be markable and giving guidance if it is not. I suggested opening a BlobOutputStream in those cases. I can also add a suggestion to consider wrapping it in a BufferedStream. I didn't add those javadocs to this method because it's in implementation, so those don't get generated anyway, but I can add that for our own internal purposes. 

Before calling this method, the calling method should be calling mark to ensure it always resets to the same place, although I'm thinking I should actually move that call to be within this method. 

I'm not terribly concerned about FIS in this case because I think if customers are uploading file data they are going to be calling uploadFromFile.

It's also worth noting that requiring markability is consistent with the expectations we make on the fluxes passed in to the async equivalent--we require that those be replayable

Reply:
Don't the async operations support non-replayable publishers? This [javadoc](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java#L242) says that the flux doesn't have to be replayable.

Reply:
Those are on the BlobClient, which does support non-replayable publishers for both async and sync. The analogues I'm referring to are on BlockBlobAsyncClient. e.g. [stage block](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlockBlobAsyncClient.java#L229)